### ISSUE TITLE
[JENKINS-28968] Aborting builds does not kill surefire sub-process

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-06-a2#708]
<p>I have a test that (unfortunately) occasionally hangs waiting on an external dependency.  I recently noticed that if the test is aborted, the surefire instance remains running on the slave machine!</p>

<p>This does not happen when running "sleep 50000" in a command window (i.e. this is killed with the job).</p>

**Attachments:**

- [JENKINS-28968.txt](https://issues.jenkins.io/secure/attachment/29985/JENKINS-28968.txt)

---
<details><summary><i>Originally reported by <a title="Ryan Desmond" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rddesmond">rddesmond</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-28968" target="_blank">Aborting builds does not kill surefire sub-process</a></i></summary>
<i><ul>
<li><b>reported by</b>: rddesmond
<li><b>status</b>: Reopened
<li><b>priority</b>: Major
<li><b>resolution</b>: Unresolved
<li><b>votes</b>: 6
<li><b>watchers</b>: 15
<li><b>imported</b>: 2025-11-06
</ul></i>
</details>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=danielbeck">danielbeck</a>:</i>
<p>Is this a freestyle project with Maven build step, or a Maven project?</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rddesmond">rddesmond</a>:</i>
<p>I worked out a baseline procedure to recreate the problem.</p>

<p>Steps:<br/>
 1. Create a new Maven Project<br/>
 2. Add a shell pre-step to setup the workspace.  It should have the contents of <span class="nobr"><a href="https://issues.jenkins.io/secure/attachment/29985/29985_JENKINS-28968.txt" title="JENKINS-28968.txt attached to JENKINS-28968">JENKINS-28968.txt<sup><img class="rendericon" src="https://issues.jenkins.io/images/icons/link_attachment_7.gif" height="7" width="7" align="absmiddle" alt="" border="0"/></sup></a></span><br/>
 3. Add the goal "test" to the build step<br/>
 4. Run and after the console prints "Now sleeping" Abort.</p>

<p>Expected Results:<br/>
 1. Surefire is no longer running on the slave machine</p>

<p>Actual Results:<br/>
 1. Surefire remains running</p>

<blockquote>
<p>$ ps aux | grep sure<br/>
<span class="error">&#91;user&#93;</span>   4222  4.7  0.1 6102248 31612 ?       Sl   10:12   0:00 java -jar /home/<span class="error">&#91;user&#93;</span>/jenkins/workspace/sleeptest/target/surefire/surefirebooter449566822541979931.jar /home/<span class="error">&#91;user&#93;</span>/jenkins/workspace/sleeptest/target/surefire/surefire7684357083774633779tmp /home/<span class="error">&#91;user&#93;</span>/jenkins/workspace/sleeptest/target/surefire/surefire_01856690741005869733tmp</p></blockquote>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rddesmond">rddesmond</a>:</i>
<p>I did notice that it killed the immediate subprocess.</p>

<p>Just before aborting:</p>
<blockquote>
<p>$ ps aux | grep sure<br/>
<span class="error">&#91;user&#93;</span>   4220  0.0  0.0 113120  1188 ?S    10:12   0:00 /bin/sh -c cd /home/ussuser/jenkins/workspace/sleeptest && /usr/lib/jvm/java-1.7.0-openjdk-1.7.0.51-2.4.5.5.el7.x86_64/jre/bin/java -jar /home/<span class="error">&#91;user&#93;</span>/jenkins/workspace/sleeptest/target/surefire/surefirebooter449566822541979931.jar /home/<span class="error">&#91;user&#93;</span>/jenkins/workspace/sleeptest/target/surefire/surefire7684357083774633779tmp /home/<span class="error">&#91;user&#93;</span>/jenkins/workspace/sleeptest/target/surefire/surefire_01856690741005869733tmp<br/>
<span class="error">&#91;user&#93;</span>   4222 13.0  0.1 6102248 31668 ?       Sl   10:12   0:00 java -jar /home/<span class="error">&#91;user&#93;</span>/jenkins/workspace/sleeptest/target/surefire/surefirebooter449566822541979931.jar /home/<span class="error">&#91;user&#93;</span>/jenkins/workspace/sleeptest/target/surefire/surefire7684357083774633779tmp /home/<span class="error">&#91;user&#93;</span>/jenkins/workspace/sleeptest/target/surefire/surefire_01856690741005869733tmp</p></blockquote>

<p>After aborting:</p>
<blockquote>
<p>$ ps aux | grep sure<br/>
<span class="error">&#91;user&#93;</span>   4222  5.2  0.1 6102248 31612 ?       Sl   10:12   0:00 java -jar /home/<span class="error">&#91;user&#93;</span>/jenkins/workspace/sleeptest/target/surefire/surefirebooter449566822541979931.jar /home/<span class="error">&#91;user&#93;</span>/jenkins/workspace/sleeptest/target/surefire/surefire7684357083774633779tmp /home/<span class="error">&#91;user&#93;</span>/jenkins/workspace/sleeptest/target/surefire/surefire_01856690741005869733tmp</p></blockquote>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rddesmond">rddesmond</a>:</i>
<p>I think this is a duplicate of <a href="https://issues.jenkins.io/browse/JENKINS-26048" title="Jenkins no longer cleaning up child processes when build stopped - as of 1.587" class="issue-link" data-issue-key="JENKINS-26048">JENKINS-26048</a>.  Closing and moving the conversation there.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rddesmond">rddesmond</a>:</i>
<p>I think this is a duplicate, closing and moving the conversation there.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cobexer">cobexer</a>:</i>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick" class="user-hover" rel="jglick">jglick</a> Are pipelines also using this code to cleanup runaway child processes? I just had a screen session escape with a VM from a pipeline job.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=oleg_nenashev">oleg_nenashev</a>:</i>
<p>Reopening this ticket according to the discussion in <a href="https://issues.jenkins.io/browse/JENKINS-26048" title="Jenkins no longer cleaning up child processes when build stopped - as of 1.587" class="issue-link" data-issue-key="JENKINS-26048">JENKINS-26048</a> . We need to confirm that this is a generic issue AND not a Maven Project specifics.</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>Don Bogardus added a comment - 11 hours ago - edited
Verified this is still happening in latest jenkins version on CentOS 6. 
For the test I'm running java tests with the failsafe plugin, which launches multiple jvms for concurrent execution and many phantomjs processes. 
1.586 - When build is stopped, all child processes are stopped
1.587 - When build is stopped, all child processes keep running
2.66.1 - When build is stopped, all child processes keep running


Daniel Beck added a comment - 11 hours ago
Don Bogardus Are you able to reproduce this issue with a simpler environment? What are the complete steps to reproduce from scratch?
When you write 2.66.1, do you mean 2.46.1?


Don Bogardus added a comment - 10 hours ago - edited
Daniel Beck , I used the latest weekly version - 2.66.1 . (It's just what came down from yum)
I am able to reproduce it with the simple java/mvn/surefire scenario created by Ryan Desmond :
https://issues.jenkins-ci.org/browse/JENKINS-28968?focusedCommentId=230600&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-230600
With this small example, 1.586 cleans up the surefire process, and 2.66.1 does not. </pre>
</div></div>
<p> </p>

<p> </p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=oleg_nenashev">oleg_nenashev</a>:</i>
<p>CC <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=aheritier" class="user-hover" rel="aheritier">aheritier</a> just in case he knows the cause</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=aheritier">aheritier</a>:</i>
<p>I think <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=tibor17" class="user-hover" rel="tibor17">tibor17</a> worked on such issues in surefire recently</p>

<p>On windows we have this bug <a href="https://issues.apache.org/jira/browse/SUREFIRE-1261" class="external-link" target="_blank" rel="nofollow noopener">https://issues.apache.org/jira/browse/SUREFIRE-1261</a> but I think I saw several issues related to surefirebooter too</p>

<p>In Jenkins side, I never succeeded to reproduce/prove it but I'm ~ sure that this case happens when we have a slave disconnection (especially with old remoting)</p>

<p> </p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dbogardus">dbogardus</a>:</i>
<p>Just to make sure this is part of the discussion. The problem has come and gone in previous jenkins versions. I installed and tested many versions of jenkins - </p>

<p>First noticed in 1.553 </p>

<p>Fixed in version 1.565.2 (Bug - <a href="https://issues.jenkins-ci.org/browse/JENKINS-22641" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-22641</a>)</p>

<p>Came back in 1.587, and is still present in weekly 2.66.1 </p>

<p>(Was tested in on master and slaves, same results)</p>

<p>All of these tests were for the maven/surefire scenario. I tried a few things to reproduce the bug outside of this scenario and was not able to. </p>

<p> </p>
